OSDOCS-6639 External DNS with shared VPC (#68016)
openshift/openshift-docs: modules/nw-control-dns-records-public-aws-with-VPC.adoc

.Procedure

. Get the Role ARN of the IAM Role that you created to allow Account B to access Account A's Route 53 hosted zone by running the following command:
+
[source,terminal]
----
$ aws --profile account-a iam get-role --role-name user-rol1 | head -1
----
+
.Example output
+
[source,terminal]
----
ROLE	arn:aws:iam::1234567890123:role/user-rol1	2023-09-14T17:21:54+00:00	3600	/	AROA3SGB2ZRKRT5NISNJN	user-rol1
----

. Locate the private hosted zone to use with Account A's credentials by running the following command:
+
[source,terminal]
----
$ aws --profile account-a route53 list-hosted-zones | grep testextdnsoperator.apacshift.support
----
+
.Example output
+
[source,terminal]
----
HOSTEDZONES	terraform	/hostedzone/Z02355203TNN1XXXX1J6O	testextdnsoperator.apacshift.support. 5
----

. Create the `ExternalDNS` object by running the following command:
+
[source,terminal]
----
$ cat <<EOF | oc create -f -
apiVersion: externaldns.olm.openshift.io/v1beta1
kind: ExternalDNS
metadata:
  name: sample-aws 
spec:
  domains:
  - filterType: Include   
    matchType: Exact   
    name: testextdnsoperator.apacshift.support 
  provider:
    type: AWS
    aws: 

>       assumeRole: arn: arn:aws:iam::12345678901234:role/user-rol1 <1>

Reviewer: This needs a new line:
```suggestion
      assumeRole:
        arn: arn:aws:iam::12345678901234:role/user-rol1 <1>
```

  source:  
    type: OpenShiftRoute 
    openshiftRouteOptions:
      routerName: default 
EOF
----
<1> Specify the Role ARN  to have DNS records created in Account A.

. Check the records created for OCP routes by using the following command:
+
[source,terminal]
----

> $ aws --profile account-a route53 list-resource-record-sets --hosted-zone-id Z02355203TNN1XXXX1J6O --query "ResourceRecordSets[?Type == 'CNAME']" | grep console

Reviewer: *nit* This is a very minor detail, but since we are providing example output, I feel like we should be more precise for finding a single route to avoid unnecessary output. This command currently returns:
```
console-openshift-console.apps.gspence-demovpc2.devcluster.openshift.com.	300	CNAME
downloads-openshift-console.apps.gspence-demovpc2.devcluster.openshift.com.	300	CNAME
```

So my suggestion is just simply to add "console-openshift-console" so it just gets the console DNS name and not the download DNS name:
```suggestion
$ aws --profile account-a route53 list-resource-record-sets --hosted-zone-id Z02355203TNN1XXXX1J6O --query "ResourceRecordSets[?Type == 'CNAME']" | grep console-openshift-console
```

Author: done

.Procedure

. Get the Role ARN of the IAM Role that you created to allow Account B to access Account A's Route 53 hosted zone by running the following command:
+
[source,terminal]
----
$ aws --profile account-a iam get-role --role-name user-rol1 | head -1
----
+
.Example output
+
[source,terminal]
----
ROLE	arn:aws:iam::1234567890123:role/user-rol1	2023-09-14T17:21:54+00:00	3600	/	AROA3SGB2ZRKRT5NISNJN	user-rol1
----

. Locate the private hosted zone to use with Account A's credentials by running the following command:
+
[source,terminal]
----
$ aws --profile account-a route53 list-hosted-zones | grep testextdnsoperator.apacshift.support
----
+
.Example output
+
[source,terminal]
----
HOSTEDZONES	terraform	/hostedzone/Z02355203TNN1XXXX1J6O	testextdnsoperator.apacshift.support. 5
----

. Create the `ExternalDNS` object by running the following command:
+
[source,terminal]
----
$ cat <<EOF | oc create -f -
apiVersion: externaldns.olm.openshift.io/v1beta1
kind: ExternalDNS
metadata:
  name: sample-aws 
spec:
  domains:
  - filterType: Include   
    matchType: Exact   
    name: testextdnsoperator.apacshift.support 
  provider:
    type: AWS
    aws: 

>       assumeRole: arn:aws:iam::12345678901234:role/user-rol1 <1>

Reviewer: I completely missed an update to the API structure, sorry:
```suggestion
      assumeRole: 
        arn: arn:aws:iam::12345678901234:role/user-rol1 <1>
```

Author: updated

>       assumeRole: arn:aws:iam::12345678901234:role/user-rol1 <1>

Reviewer 2: assumeRole:
        arn: arn:aws:iam::123456789012:role/role-name # Replace with the desire Role ARN
        
        the entire ARN needs to be replaced

Author: @gcs278 Does this need to change as well? I don't understand this comment.

Reviewer: Everything looks okay to me - @melvinjoseph86 let us know what you were thinking by this block of code. I think the `<1>` leads to the footnote you might be implying it needs.

Reviewer 2: I meant the whole `arn:aws:iam::123456789012:role/role-name` including the `123456789012:role/role-name` will be replaced as ARN, not only the `role/user-rol1 `. 
But that is not a big deal. You can moved ahead with current changes.
/lgtm from our side

>       credentials:

Reviewer: ```suggestion
```

Whoops - I left this in the draft, and it doesn't need to be in there due to a recent fix in the API.

Author: Thanks!

  source:  
    type: OpenShiftRoute 
    openshiftRouteOptions:
      routerName: default 
EOF
----
<1> Specify the Role ARN  to have DNS records created in Account A.

> . Check the records created for OCP routes by using the following command:

Reviewer 3: OCP needs to be spelled out per [Supplementary Style Guidelines](https://redhat-documentation.github.io/supplementary-style-guide/#_r).

Author: @kcarmichael08 thank you, will do right now!

+
[source,terminal]
----

> $ aws --profile account-a route53 list-resource-record-sets --hosted-zone-id Z02355203TNN1XXXX1J6O --query "ResourceRecordSets[?Type == 'CNAME']" | grep console+

Reviewer: Did a `+` get added accidentally?  
```suggestion
$ aws --profile account-a route53 list-resource-record-sets --hosted-zone-id Z02355203TNN1XXXX1J6O --query "ResourceRecordSets[?Type == 'CNAME']" | grep console
```

Author: Oops, yes, my apologies

// Module included in the following assemblies:
//
// * networking/external_dns_operator/nw-creating-dns-records-on-aws.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-control-dns-records-public-aws-with-VPC_{context}"]
= Creating DNS records in a different AWS Account using a shared VPC

> You can use ExternalDNS Operator to create DNS records in a different AWS account using a shared Virtual Private Cloud (VPC). A shared VPC enables an organization to connect resources from multiple projects to a common VPC network. VPC sharing, in turn, enables the use of a single Route 53 instance across multiple AWS accounts.

Author: Thank you so much. Updated.

> With VPC sharing, we will have two Amazon AWS accounts, one with a VPC and a Route 53 hosted zone configured (Account A), and the other for installing a cluster (Account B). We will configure ExternalDNS Operator in Account B to create records in a Route 53 hosted zone Account A. In this example, we will illustrate the process of sharing a private hosted zone.

Author: I was thinking it could be eliminated altogether as it seems redundant with example titles below.

.Prerequisites

> * You have two Amazon AWS accounts, one with a VPC and a Route 53 private hosted zone configured (Account A), and the other for installing a cluster (Account B).

Author: Thank you

> * You have created an IAM Policy and IAM Role with the appropriate permission in Account A for Account B to create DNS records in Account A's Route 53.

Author: Thank you

> * You have installed a cluster in Account B into the existing VPC of Account A as documented in xref:../../installing/installing_aws/installing-aws-vpc.adoc#installing-aws-vpc[Installing a cluster on AWS into an existing VPC].

Author: Gotcha.

> * You have installed the ExternalDNS Operator in the cluster in Account B as documented in xref:../../networking/external_dns_operator/nw-installing-external-dns-operator-on-cloud-providers.adoc#installing-external-dns-on-cloud-providers[Installing External DNS Operator on cloud providers].

Author: Thank you

.Procedure

> . Get the Role ARN of the IAM Role you created for allowing Account B to access Account A's Route 53:

Author: Thank you.

+
[source,terminal]
----
$ aws --profile account-a iam get-role --role-name user-rol1 | head -1
----
+
.Example output
+
[source,terminal]
----
ROLE	arn:aws:iam::1234567890123:role/user-rol1	2023-09-14T17:21:54+00:00	3600	/	AROA3SGB2ZRKRT5NISNJN	user-rol1
----

> . Locate the private hosted zone you wish to use with Account A's credentials:

Author: Lovely, thank you

+
[source,terminal]
----
$ aws --profile account-a route53 list-hosted-zones | grep testextdnsoperator.apacshift.support
----
+
.Example output
+
[source,terminal]
----
HOSTEDZONES	terraform	/hostedzone/Z02355203TNN1XXXX1J6O	testextdnsoperator.apacshift.support. 5
----

> . Create the ExternalDNS object:

Reviewer 4: ```suggestion
. Create the `ExternalDNS` object by running the following command:
```
- Enclose literal names in back ticks.
- Add "by running the following command".

Author: TY

+
[source,terminal]
----
$ cat <<EOF | oc create -f -
apiVersion: externaldns.olm.openshift.io/v1beta1
kind: ExternalDNS
metadata:
  name: sample-aws 
spec:
  domains:
  - filterType: Include   
    matchType: Exact   
    name: testextdnsoperator.apacshift.support 
  provider:
    type: AWS
    aws: 
      assumeRole: arn:aws:iam::12345678901234:role/user-rol1 <1>
      credentials:
  source:  
    type: OpenShiftRoute 
    openshiftRouteOptions:
      routerName: default 
EOF
----

> <1> Specify the Role ARN  to have DNS records created in Account A.

Reviewer 4: ```suggestion
<1> Specify the Role ARN to have DNS records created in Account A.
```
- Does this mean "Specify the Role ARN so that DNS records will be created in Account A"? or does it mean something else? It's not 100% clear to me what specifying the Role ARN does. Maybe double-check with your SME or QE?

Author: @lihongan please let me know

Reviewer 3: I think there are still 2 spaces after "Role ARN" per Brian's comment.

> . Check the records created for OCP routes by using the following command:

Reviewer 4: ```suggestion
. Check that the records have been created for {product-title} routes by running the following command:
```
- Consider rewording to clarify that the user is verifying that the records have been created.
- Use an attribute for OpenShift Container Platform
- Change `using` to `running` per the OCP docs repo guidelines for introducing commands.

+
[source,terminal]
----
$ aws --profile account-a route53 list-resource-record-sets --hosted-zone-id Z02355203TNN1XXXX1J6O --query "ResourceRecordSets[?Type == 'CNAME']" | grep console+

> ----

Reviewer 4: I recommend including example output following the verification command, if possible.

Author: @gcs278 do you have example output?

Reviewer: Yes - I got it now, if you make the change I suggested in https://github.com/openshift/openshift-docs/pull/68016/files#r1399775161, it should be:
```
console-openshift-console.apps.gspence-demovpc2.devcluster.openshift.com.	300	CNAME
```

Reviewer: Oh, one more minor point - if we add example output, should we be consistent by also adding the same example output here:
https://github.com/openshift/openshift-docs/blob/3770f73e0051c332dd18cb45f911e1e07a10e0d5/modules/nw-control-dns-records-public-hosted-zone-aws.adoc?plain=1#L96

This is the existing module that will be right above this new one (https://docs.openshift.com/container-platform/4.13/networking/external_dns_operator/nw-creating-dns-records-on-aws.html). I used this module as a template (hence why I didn't have example output). Might be good to be consistent since they are so similar.
